src/storage/cache/rpc/packet.rs

```diff
@@ -0,0 +1,524 @@
+                        // Set the task as timeout
+                        debug!("Task {} is timeout", seq);
+                        packet.set_status(PacketStatus::Timeout);
+                    }
+                }
+            }
+        }
+
+        // clean the timeout packets
+        for seq in last_timeout_packets {
+            self.packets.remove(&seq);
+            self.timestamp.remove(&seq);
+        }
+    }
+
+    /// Get task from the packets
+    pub fn get_task_mut(&mut self, seq: u64) -> Option<(&mut P, u64)> {
+        if let Some(packet) = self.packets.get_mut(&seq) {
+            if let Some(timestamp) = self.timestamp.get(&seq) {
+                return Some((packet, *timestamp));
+            }
+
+            // Do not need to remove the packet if the timestamp is not found
+            return None;
+        }
+
+        None
+    }
+}
+
+/// The `PacketsKeeper` struct is used to store the current and previous tasks.
+/// It will be modified by single task, so we don't need to share it.
+#[derive(Debug)]
+pub struct PacketsKeeper<P>
+where
+    P: Packet + Send + Sync,
+{
+    /// current tasks, marked by the seq number
+    inner: Arc<RwLock<PacketsInner<P>>>,
+    /// buffer sender
+    buffer_packets_sender: flume::Sender<(u64, P)>,
+    /// buffer receiver
+    buffer_packets_receiver: flume::Receiver<(u64, P)>,
+    /// The maximum number of tasks that can be stored in the previous_tasks
+    /// We will mark the task as timeout if it is in the previous_tasks and the previous_tasks is full
+    timeout: u64,
+    /// current timestamp
+    current_time: Instant,
+}
+
+impl<P: Packet + Send + Sync> PacketsKeeper<P> {
+    /// Create a new `PacketsKeeper`
+    #[must_use]
+    pub fn new(timeout: u64) -> Self {
+        let (buffer_packets_sender, buffer_packets_receiver) = flume::bounded::<(u64, P)>(1000);
+        let packets_inner = Arc::new(RwLock::new(PacketsInner::new()));
+        let current_time = tokio::time::Instant::now();
+
+        PacketsKeeper {
+            inner: packets_inner,
+            buffer_packets_sender,
+            buffer_packets_receiver,
+            timeout,
+            current_time,
+        }
+    }
+
+    /// Add a task to the packets
+    pub fn add_task(&mut self, packet: P) -> Result<(), RpcError<String>> {
+        // TODO: use a global atomic ticker(updated by check_loop) or read current time?
+        let timestamp = self.current_time.elapsed().as_secs();
+        self.buffer_packets_sender
+            .send((timestamp, packet))
+            .map_err(|e| {
+                RpcError::InternalError(format!("Failed to send packet to buffer: {e:?}"))
+            })?;
+
+        Ok(())
+    }
+
+    /// Consume a task from inner
+    /// If return task is none, means that current task is in buffer,
+    /// and haven't triggered timeout or get response in time
+    pub fn consume_task(&mut self, seq: u64) -> Option<P> {
```

Comment:
Unused function, delete it,  (duplicated of update_task_mut)

Reply:
OK.